### PR TITLE
Shane image

### PR DESCRIPTION
**Pull request by MTN718**
Added setImage method for Web,Android,IOS Platform
Parameter is Image URL

### CODE REVIEW

File: android/src/main/java/com/reactnativecommunity/clipboard/ClipboardModule.java
Comment (roryabraham) [inline android/src/main/java/com/reactnativecommunity/clipboard/ClipboardModule.java]: Not sure this comment is useful

File: android/src/main/java/com/reactnativecommunity/clipboard/ClipboardModule.java
Comment (roryabraham) [inline android/src/main/java/com/reactnativecommunity/clipboard/ClipboardModule.java]:
```suggestion
    File file = Environment.getExternalStorageDirectory();
```

File: android/src/main/java/com/reactnativecommunity/clipboard/ClipboardModule.java
Comment (roryabraham) [inline android/src/main/java/com/reactnativecommunity/clipboard/ClipboardModule.java]:
```suggestion
      ContentResolver contentResolver = getContext().getContentResolver();
```

File: android/src/main/java/com/reactnativecommunity/clipboard/ClipboardModule.java
Comment (roryabraham) [inline android/src/main/java/com/reactnativecommunity/clipboard/ClipboardModule.java]:
```suggestion
      ClipData clipData = ClipData.newUri(theContent, dest.getName(), imageUri);
```

File: src/Clipboard.ts
Comment (roryabraham) [inline src/Clipboard.ts]: Missing JSDoc

File: src/Clipboard.web.ts
Comment (roryabraham) [inline src/Clipboard.web.ts]: Missing JSDoc

File: src/Clipboard.ts
Comment (roryabraham) [inline src/Clipboard.ts]:
```suggestion
      .then((response: Response) => {
```

File: src/Clipboard.ts
Comment (roryabraham) [inline src/Clipboard.ts]:
```suggestion
      .then(blob: Blob => new Promise(() => {
```

`callback` seems unused.

File: src/Clipboard.ts
Comment (roryabraham) [inline src/Clipboard.ts]:
```suggestion
          let reader = new FileReader();
```

File: src/Clipboard.ts
Comment (roryabraham) [inline src/Clipboard.ts]: This is a bit confusing – you're using a regular function an not an arrow function because you want `this` to refer to the `FileReader`?

I might change this like so:

```javascript
reader.onload = () => {
    let base64Data = reader.result.replace("image/jpeg", "image/png");
    NativeClipboard.setImage(base64Data);
}
```

File: src/Clipboard.ts
Comment (roryabraham) [inline src/Clipboard.ts]: Probably want to return the promise from this function:

```suggestion
    return fetch(imageUrl)
```

File: src/Clipboard.ts
Comment (roryabraham) [inline src/Clipboard.ts]: Probably want to remove this `catch` block as it's not doing anything right now.

File: src/Clipboard.web.ts
Comment (roryabraham) [inline src/Clipboard.web.ts]:
```suggestion
    write(data:Array<any>): Promise<void>;
```

File: src/Clipboard.web.ts
Comment (roryabraham) [reply src/Clipboard.web.ts]: with some more types and using purely async/await:

```javascript
async function setImage(imageURL: string): Promise<void> {
	if (navigator && navigator.clipboard) {
	      const response: Response = await fetch(imageURL);
	      if (!response.ok) { throw Error(response.statusText); }
	      const blob: Blob = await response.blob();
	      const reader = new FileReader();
	      reader.onload = async () => {
	          let base64Data = this.result as string;
	          base64Data = base64Data.replace("image/jpeg", "image/png");
	          const base64Response: Response = await fetch(base64Data);
	          const innerBlob: Blob = await base64Response.blob();
	          await navigator.clipboard.write([new ClipboardItem({"image/png": innerBlob})]);
	      }
	      await reader.readAsDataURL(blob);
	}
}
```

File: tsconfig.json
Comment (roryabraham) [inline tsconfig.json:14]: I'm somewhat doubtful that the upstream maintainers will be okay with enabling `skipLibCheck`

**Review by roryabraham — CHANGES_REQUESTED** [review]

**Comment by roryabraham** [comment]
I need a second reviewer from one of our more engineers more experienced in native mobile development. Unfortunately I'm very unfamiliar with Objective-C

**Comment by roryabraham** [comment]
@Expensify/mobile-deployers Any volunteers please? I need a second opinion on this PR. It's related to https://github.com/Expensify/App/issues/4629

**Comment by MTN718** [comment]
Ok let me fix following your feedback.

File: android/src/main/java/com/reactnativecommunity/clipboard/ClipboardModule.java
Comment (MTN718) [reply android/src/main/java/com/reactnativecommunity/clipboard/ClipboardModule.java]: To decode base64 in android, need to remove prefix.

File: src/Clipboard.ts
Comment (MTN718) [reply src/Clipboard.ts]: not sure this one, can you explain better?

**Review by Julesssss — CHANGES_REQUESTED** [review]
Sorry for the late review here. I think we should continue with the Web/Desktop/mWeb copy image logic, but remove the Android/iOS clipboard copy functionality. We shouldn't port this desktop feature to mobile. 

Sharing images on both mobile platforms should be done via their native share functionality ([Android](https://developer.android.com/training/sharing), [iOS](https://developer.apple.com/design/human-interface-guidelines/ios/extensions/sharing-and-actions/)). This allows the user to share images to any app which has registered their support of receiving images. 

The one case that this might be considered useful is for sharing internally within our app, but again we should not depend on copy/paste here. We should instead register our app as a receiver of images (native share), which allows any other app to share images directly into our chats. If we are desperate to add this functionality right now then instead we can build our own UI for sharing images to a different chat.

**Review by Jag96 — COMMENTED** [review]
@MTN718 can you please update the issue description to include the issue link and your testing steps?

**Comment by Julesssss** [comment]
Update to my review: @roryabraham let me know that this is actually common on iOS, so I take back that part of my argument. I still strongly feel that it's not worth hacking an Android solution purely for feature parity here.

But if nobody else agrees then let's just include it 🤷

File: android/src/main/java/com/reactnativecommunity/clipboard/ClipboardModule.java
Comment (Julesssss) [reply android/src/main/java/com/reactnativecommunity/clipboard/ClipboardModule.java]: I think the point is that the variable is already named `Bitmap`, so the comment adds nothing here.

File: android/src/main/java/com/reactnativecommunity/clipboard/ClipboardModule.java
Comment (Julesssss) [inline android/src/main/java/com/reactnativecommunity/clipboard/ClipboardModule.java:129]: We probably shouldn't silently fail here. Can we callback to the JS layer that the copy failed?

File: ios/NSData+Base64.m
Comment (Julesssss) [inline ios/NSData+Base64.m:40]: 😬

File: src/Clipboard.ts
Comment (Julesssss) [reply src/Clipboard.ts]: Unrelated, but what is a `.ts` file? Is that for typescript or something?

File: src/Clipboard.web.ts
Comment (Julesssss) [inline src/Clipboard.web.ts]: Useless lines. Can we handle the error properly?

File: src/Clipboard.ts
Comment (roryabraham) [reply src/Clipboard.ts]: Yep, TypeScript.

File: android/src/main/java/com/reactnativecommunity/clipboard/ClipboardModule.java
Comment (AndrewGable) [inline android/src/main/java/com/reactnativecommunity/clipboard/ClipboardModule.java:121]:
```suggestion
      values.put(MediaStore.Images.Media.MIME_TYPE, "image/png");
```

Can you just run google format on your entire code change? https://google.github.io/styleguide/javaguide.html

There should be a plugin for whatever editor you are using.

File: src/Clipboard.ts
Comment (roryabraham) [reply src/Clipboard.ts]: Oh, basically this is a function that performs some asynchronous actions that take some time to complete. Right now, the function isn't returning anything. But if you return a promise (even if it's a void promise), then when using this function we can wait for the image to finish copying to the clipboard _then_ do something else.

```javascript
Clipboard.setImage(myURL).then(() => doSomethingElse());
```

**Comment by AndrewGable** [comment]
The objective-c looks fine to me, but fair warning it's been a while since I've written any myself.

**Comment by roryabraham** [comment]
A question that just came to mind – we're using `fetch` to fetch the URL of an image in this code ... does that mean we're reliant on having a network connection for copying the image to the clipboard to work? Like, if we've got our app open but we're on a plane and have no internet connection, will we still be able to copy an image to the clipboard?

**Comment by MTN718** [comment]
Question: Need to remove copy image code for android?

**Comment by MTN718** [comment]
Fixed following feedback @roryabraham

**Comment by MTN718** [comment]
Can you check and give me feedback? @roryabraham

**Comment by roryabraham** [comment]
Did you have any comment on [this question](https://github.com/Expensify/clipboard/pull/1#issuecomment-912761173)?

**Comment by MTN718** [comment]
Regarding this question, can't copy clipboard on internet because currently system provide url of attachment so it should be fetch content of image from internet.

**Comment by Julesssss** [comment]
> can't copy clipboard on internet because currently system provide url of attachment so it should be fetch content of image from internet.

Hmm, that seems a bit of a problem. It makes the feature inconsistent 😕

**Comment by MTN718** [comment]
> > can't copy clipboard on internet because currently system provide url of attachment so it should be fetch content of image from internet.
> 
> Hmm, that seems a bit of a problem. It makes the feature inconsistent 😕

But to get base64 image data from image url, fetch image necessary. Or we can send parameter base64 image for that library and fetch action in the app code.

**Comment by Julesssss** [comment]
> But to get base64 image data from image url, fetch image necessary. Or we can send parameter base64 image for that library and fetch action in the app code.

Can you expand on the second part of your sentence, please? I think you are suggesting that this would work offline? Presumably the iOS apps that implement image copy/paste have solved this without requiring a connection?

**Comment by MTN718** [comment]
Regarding those ios app, it is getting UIImage from UIImageView and set  UIImage to Clipboard. But in expensify app, currently not getting UIImage from chat image, only url, so fetch base64 data from url and make UIImage and set to clipboard 

And Regarding second sentence, I mean we can fetch image base64 data in the app and send base64 image data to library for clipboard and handle base64 data in the library not image url

**Comment by MTN718** [comment]
@roryabraham Please give me feedback

**Comment by roryabraham** [comment]
Basically, we need this feature to work offline. The current approach won't work offline, and it sounds like this one won't either:

> I mean we can fetch image base64 data in the app and send base64 image data to library for clipboard and handle base64 data in the library not image url

So it sounds like what we need is something more like this:

1. This library's `setImage` function needs to take a RN image component
1. Then use the RN bridge + native code to get the base64 image data directly from the UIImage
1. Then process the base64 image and set it to the clipboard.

**Comment by MTN718** [comment]
@roryabraham Then in app, need to fetch base64 image data from url and pass that data object when we call library to copy clipboard, and handle that image data in native. Because currently message is img element with url

image

**Comment by roryabraham** [comment]
@MTN718 That solution still won't work offline – it just moved the `fetch` from inside the Clipboard library to our app.

**Comment by MTN718** [comment]
But current app's structure store message to img element so can get only image url. I think need to change sturcture for store messages

**Comment by kadiealexander** [comment]
Hi @MTN718, I've placed this issue on hold as per [this update](https://expensify.slack.com/archives/C01GTK53T8Q/p1631754482374800), as we are prioritising issues related to a feature release scheduled for 9/24. As an apology for the delay, we will add a $100 bonus as a thank you for waiting.

**Comment by MTN718** [comment]
this PR still hold?

**Comment by roryabraham** [comment]
PR is no longer on HOLD, however, we'll need a solution that works offline.

**Pull request closed without merging**
